### PR TITLE
Return 0/1 from comparison operators in Kernel

### PR DESCRIPTION
**Pull request by josevalim**
Otherwise someone can write this code:

```elixir
defn example(matrix) do
  {m, n} = Nx.shape(matrix)
  if m == n do
    ...
  else
    ...
  end
end
```

And a boolean would be given to `if`, making it raise. Alternatively we could allow booleans in `if`, `and`, etc. But I don't think we should go down this slippery slope yet. :D

### CODE REVIEW

File: nx/test/nx/defn/kernel_test.exs [hunk 1]
```diff
@@ -23,6 +23,15 @@ defmodule Nx.Defn.KernelTest do
       assert Nx.Defn.Kernel./(1, 2) == 0.5
     end
 
+    test "comparison" do
+      assert Nx.Defn.Kernel.==(0, 0) == 1
+      assert Nx.Defn.Kernel.!=(0, 0) == 0
+      assert Nx.Defn.Kernel.>(0, 0) == 0
+      assert Nx.Defn.Kernel.>=(0, 0) == 1
+      assert Nx.Defn.Kernel.<(0, 0) == 0
+      assert Nx.Defn.Kernel.<=(0, 0) == 1
```
Comment (polvalente) [inline nx/test/nx/defn/kernel_test.exs:32]: Shouldn't we return Nx.tensor(0, type: {:u, 8}) and the equivalent for 1, instead?
I think it would make for more uniform return types in defn, even though the overrides would make eveything work fine...

Approving this nevertheless

**Review by polvalente — APPROVED** [review]

File: nx/test/nx/defn/kernel_test.exs (hunk 1 above)
Comment (josevalim) [reply nx/test/nx/defn/kernel_test.exs:32]: You are correct. I will tackle this.

File: nx/lib/nx/binary_backend.ex [hunk 2]
```diff
@@ -22,8 +22,8 @@ defmodule Nx.BinaryBackend do
   ## Creation
 
   @impl true
-  def scalar(%{type: type, shape: shape} = out, scalar, _backend_options) do
-    data = :binary.copy(number_to_binary(scalar, type), Nx.size(shape))
+  def constant(%{type: type, shape: shape} = out, constant, _backend_options) do
```
Comment (polvalente) [inline nx/lib/nx/binary_backend.ex:25]: Why the change from scalar to constant?

Comment (polvalente) [reply nx/lib/nx/binary_backend.ex:25]: I think that scalar reflects better the mathematical sense, whereas constant seems to be the opposite of variable (which isn't the case)

Comment (josevalim) [reply nx/lib/nx/binary_backend.ex:25]: @polvalente it is a constant because it can have any shape. :)

Comment (polvalente) [reply nx/lib/nx/binary_backend.ex:25]: Nice, I missed that :)

Comment (josevalim) [reply nx/lib/nx/binary_backend.ex:25]: To be clear, it was previously incorrect, it was always a constant because we use it for optimizations, but I noticed it only now when exposing the API.

File: nx/lib/nx/defn/kernel.ex [hunk 3]
```diff
@@ -536,23 +536,19 @@ defmodule Nx.Defn.Kernel do
   def not tensor when is_number(tensor), do: logical_not(tensor)
   def not tensor, do: Nx.logical_not(tensor)
 
-  defp logical_and(l, r) when Kernel.==(l, 0), do: zero(l, r)
-  defp logical_and(l, r) when Kernel.==(r, 0), do: zero(l, r)
-  defp logical_and(l, r), do: one(l, r)
+  defp logical_and(l, _) when Kernel.==(l, 0), do: zero()
```
Comment (polvalente) [inline nx/lib/nx/defn/kernel.ex:539]:
```suggestion
  defp logical_and(0, _), do: zero()
```
Why not hard-match like this?

**Review by polvalente — APPROVED** [review]
Other than the last stylistic question, the code looks fine

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: nx/lib/nx/defn/kernel.ex (hunk 3 above)
Comment (josevalim) [reply nx/lib/nx/defn/kernel.ex:539]: We need to handle both floats and integers and a match is type specific. :)

Comment (polvalente) [reply nx/lib/nx/defn/kernel.ex:539]: Nice. == vs ===
